tools/bdf-to-mono/src/lib.rs

```diff
@@ -34,32 +28,106 @@ pub fn bdf_to_bitmap(font: &BdfFont, encoding: Encoding) -> Result<Bitmap> {
+    // to maintain the correct kerning.
+    let mut char_spacing = None;
+
+    // find the maximum bounding box that can fit all glyphs taking offsets into account
+    // glyph_bb.offset.y will contain the baseline
+    // additionally build the `selected_glyphs` vector to save on the glyph lookup later in the code
+    let selected_glyphs = chars
+        .into_iter()
+        .enumerate()
+        .map(|(index, c)| {
+            let glyph = font.glyphs.get(c);
+            let map = (index, c, glyph);
+            if let Some(glyph) = glyph {
+                let off_x = glyph.bounding_box.offset.x;
+                if off_x < glyph_bb.offset.x {
+                    glyph_bb.offset.x = off_x;
+                }
+                let off_y = glyph.bounding_box.offset.y;
+                if off_y < glyph_bb.offset.y {
+                    glyph_bb.offset.y = off_y;
+                }
+                if off_y > max_y_offset {
+                    max_y_offset = off_y;
+                }
+                let gl_w = glyph.bounding_box.size.x + off_x.abs();
+                if gl_w > glyph_bb.size.x {
+                    glyph_bb.size.x = gl_w;
+                }
+                let gl_h = glyph.bounding_box.size.y + off_y.abs();
+                if gl_h > glyph_bb.size.y {
+                    glyph_bb.size.y = gl_h;
+                }
+                let x_size_diff = glyph.device_width.x - glyph.bounding_box.size.x;
+                if char_spacing.map_or(true, |char_spacing_compensation| {
+                    x_size_diff < char_spacing_compensation
+                }) {
+                    char_spacing = Some(x_size_diff);
+                }
+            }
+            map
+        })
+        .collect::<Vec<_>>();
+    let glyph_width = usize::try_from(glyph_bb.size.x)?;
+    let mut glyph_height = usize::try_from(glyph_bb.size.y)?;
+    // compensate the glyph height if case all offsets are either > 0 or < 0
+    let y_offset_compensation;
+    if max_y_offset.signum() == glyph_bb.offset.y.signum() {
+        y_offset_compensation = max_y_offset.abs().min(glyph_bb.offset.y.abs());
+        glyph_height = usize::try_from(i32::try_from(glyph_height)? - y_offset_compensation)?;
+    } else {
+        y_offset_compensation = 0;
+    }
     let bitmap_width = ((glyph_width * 16 + 7) / 8) * 8;
     let bitmap_height = glyph_height * rows.len();
     let mut bitmap = vec![false; bitmap_width * bitmap_height];
 
-    for (index, c) in chars.enumerate() {
-        if let Some(glyph) = font.glyphs.iter().find(|glyph| glyph.encoding == Some(c)) {
-            ensure!(
-                glyph.bounding_box == glyph_bb,
-                "glyph '{}' has a different bounding",
-                c
-            );
+    let fallback_glyph = font
+        .glyphs
+        .get('?')
+        .ok_or_else(|| anyhow!("font doesn't contain the fallback '?' glyph"))?;
```

Comment:
I would move this into the `map` function for `selected_glyphs`. This way the question mark will be included in the bounding box calculation, if it is used. This won't make a difference at the moment, because all subsets include the `?` anyway, but if we decide other subsets in the future it could make a difference.

Reply:
Done!